File: Educative/SlidingWindow/LongestSubstringWithDistinctCharacters.py
# Given a string, find the length of the longest substring, which has all distinct characters.
# Time O(n), Space O(k) where k is the number of distinct characters in the string

def non_repeat_substring(str1):
  window_start, max_length = 0, 0
  index_map = {}

  for window_end in range(len(str1)):
    right_char = str1[window_end]
    if right_char in index_map:
      window_start = max(window_start, index_map[right_char] + 1)
    
    index_map[right_char] = window_end
    max_length = max(max_length, window_end - window_start + 1)
  return max_length

File: Educative/SlidingWindow/LongestSubstringWithDistinctCharacters.java
// Given a string, find the length of the longest substring, which has all distinct characters.
// Time O(n), Space O(k) where k is the number of distinct characters in the string

import java.util.*;

class NoRepeatSubstring {
  public static int findLength(String str) {
    int windowStart = 0, maxLength = 0;
    Map<Character, Integer> indexMap = new HashMap<>();

    for (int windowEnd = 0; windowEnd < str.length(); windowEnd++){
      char rightChar = str.charAt(windowEnd);
      if (indexMap.containsKey(rightChar))
        windowStart = Math.max(windowStart, indexMap.get(rightChar) + 1);

      indexMap.put(rightChar, windowEnd);
      maxLength = Math.max(maxLength, windowEnd - windowStart + 1);
    }
    return maxLength;
  }
}